server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeExecutor.java

```diff
@@ -0,0 +1,156 @@
+    /**
+     * Floor for IO write rate limit (we will never go any lower than this)
+     */
+    private static final double MIN_MERGE_MB_PER_SEC = 5.0;
+    /**
+     * Ceiling for IO write rate limit (we will never go any higher than this)
+     */
+    private static final double MAX_MERGE_MB_PER_SEC = 10240.0;
+    /**
+     * Initial value for IO write rate limit when doAutoIOThrottle is true
+     */
+    private static final double START_MB_PER_SEC = 20.0;
+    /**
+     * Current IO write throttle rate, for all merge, across all merge schedulers (shards) on the node
+     */
+    private double targetMBPerSec = START_MB_PER_SEC;
+    private final SortedSet<ThreadPoolMergeScheduler> registeredMergeSchedulers = new TreeSet<>(new Comparator<ThreadPoolMergeScheduler>() {
+        @Override
+        public int compare(ThreadPoolMergeScheduler tpms1, ThreadPoolMergeScheduler tpms2) {
+            MergeTask mergeTask1 = tpms1.peekMergeTaskToExecute();
+            MergeTask mergeTask2 = tpms2.peekMergeTaskToExecute();
+            if (mergeTask1 == null && mergeTask2 == null) {
+                // arbitrary order between schedulers that cannot run any merge right now
+                return System.identityHashCode(mergeTask1) - System.identityHashCode(mergeTask2);
+            } else if (mergeTask1 == null) {
+                // "merge task 2" can run because "merge scheduler 1" cannot run any merges
+                return 1;
+            } else if (mergeTask2 == null) {
+                // "merge task 1" can run because "merge scheduler 2" cannot run any merges
+                return -1;
+            } else {
+                // run smaller merge task first
+                return mergeTask1.compareTo(mergeTask2);
+            }
+        }
+    });
+    private final ExecutorService executorService;
+    private final int maxConcurrentMerges;
+    private int currentlyExecutingMergesCount;
+    private int currentlyActiveIOThrottledMergesCount;
+
+    public ThreadPoolMergeExecutor(ThreadPool threadPool) {
+        this.executorService = threadPool.executor(ThreadPool.Names.MERGE);
+        this.maxConcurrentMerges = threadPool.info(ThreadPool.Names.MERGE).getMax();
+    }
+
+    public double getTargetMBPerSec() {
+        return targetMBPerSec;
+    }
+
+    public synchronized void registerMergeScheduler(ThreadPoolMergeScheduler threadPoolMergeScheduler) {
+        if (registeredMergeSchedulers.add(threadPoolMergeScheduler) == false) {
+            throw new IllegalStateException("cannot register the same scheduler multiple times");
+        }
+    }
+
+    public synchronized void unregisterMergeScheduler(ThreadPoolMergeScheduler threadPoolMergeScheduler) {
+        if (registeredMergeSchedulers.remove(threadPoolMergeScheduler) == false) {
+            throw new IllegalStateException("cannot unregister if the scheduler has not been registered");
+        }
+    }
+
+    public synchronized void updateMergeScheduler(
```

Comment:
I have some perhaps unqualified anxiety around this central mutex across the node. Probably not a real issue but noting it anyway.

Reply:
Yeah, this is an issue, but I don't see a better way in the current design.